proxy/src/registry/transaction.rs

```diff
@@ -0,0 +1,338 @@
+//! Abstractions and types to handle, persist and interact with transactions.
+
+use async_trait::async_trait;
+use hex::ToHex;
+use kv::Codec as _;
+use serde::{Deserialize, Serialize};
+use std::time::SystemTime;
+
+use radicle_registry_client as protocol;
+
+use crate::error;
+use crate::registry;
+
+/// A container to dissiminate and apply operations on the [`Registry`].
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Transaction {
+    /// Unique identifier, in actuality the Hash of the transaction. This handle should be used by
+    /// the API consumer to query state changes of a transaction.
+    pub id: registry::Hash,
+    /// List of operations to be applied to the Registry. Currently limited to exactly one. We use
+    /// a Vec here to accommodate future extensibility.
+    pub messages: Vec<Message>,
+    /// Current state of the transaction.
+    pub state: State,
+    /// Creation time.
+    pub timestamp: SystemTime,
+}
+
+/// Possible messages a [`Transaction`] can carry.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase", tag = "type")]
+pub enum Message {
+    /// Issue a new org registration.
+    OrgRegistration {
+        /// The [`registry::Org`] id.
+        id: registry::Id,
+    },
+
+    /// Issue an org unregistration with a given id.
+    OrgUnregistration {
+        /// The [`registry::Org`] id.
+        id: registry::Id,
+    },
+
+    /// Issue a new project registration with a given name under a given org.
+    ProjectRegistration {
+        /// Actual project name, unique for org.
+        project_name: registry::ProjectName,
+        /// The Org in which to register the project.
+        org_id: registry::Id,
+    },
+
+    /// Issue a user registration for a given handle storing the corresponding identity id.
+    UserRegistration {
+        /// Globally unique user handle.
+        handle: registry::Id,
+        /// Identity id originated from librad.
+        id: Option<String>,
```

Comment:
Kind of expect this to be binary data, ie. `Vec<u8>`. Definitely not a utf8 string :thinking: